### DISCUSSION TITLE
Importing Custom JSON File

### DISCUSSION THREAD

**Discussion by Nebula-Ninjas** [OpenCSPM/opencspm#152] [Q&A] [answered]
Dear OpenCSPM,

I am rather new to this Docker stuff, and my question may be just my lack of knowledge around the platform. In short, when I create the custom JSON file, extract to the folder specified for OpenCSPM to import. I then attempt to reload the data _(via web console / docker-compose down/up)_ I get the following error.

worker_1  | Warning: no Backup loader defined for asset type: protected_resource
worker_1  | Loader job failed - NameError uninitialized constant AWSLoader::ApplicationAutoScaling (/app/app/jobs/lib/loader/asset/loaders/aws_loader/_resource_loader.rb:25:in `const_get')
worker_1  | Runner job failed - NameError uninitialized constant AWSLoader::ApplicationAutoScaling (/app/app/jobs/lib/loader/asset/loaders/aws_loader/_resource_loader.rb:25:in `const_get')
worker_1  | I, [2021-04-21T14:06:28.283442 #1]  INFO -- : [ActiveJob] [RunnerJob] [f82dc9b3-19f5-485b-b664-f60a937113ed] Performed RunnerJob (Job ID: f82dc9b3-19f5-485b-b664-f60a937113ed) from Sidekiq(default) in 44876.54ms
worker_1  | 2021-04-21T14:06:28.295Z pid=1 tid=gp5k0oly1 class=RunnerJob jid=535e242bd81a6098d34bbb62 elapsed=44.947 INFO: done

I hope this is just my lack of knowledge around the dependencies of the tool and you all can provide some insight as to remedy. 

On another note, tool is AWESOME and one of the easiest to get up and working from other tools I've used. Thank you all for creating this and hopefully my issue is easily resolve with a few commands on my part.

Thanks in advance!

William

**Comment by joshlarsen** [comment, marked as answer]
Hi William. This actually a bug. We've opened [issue 153](https://github.com/OpenCSPM/opencspm/issues/153) for it. We'll work on a fix shortly.

Thanks!

**Reply by Nebula-Ninjas** [reply]
Thanks, @joshlarsen 

I look forward to the fix and getting it up and running.

Sincerely,

William

**Reply by joshlarsen** [reply]
William,

Please try the following:

bring down the stack:

`docker-compose down`

update the `0.1.20` image:

`docker pull gcr.io/opencspm/engine:0.1.20`

bring the stack back up:

`docker-compose up`

Note, the updated image digest is: `sha256:8d7fc48cde410547d5bb148b96aa104ebed1d4dfcd83e578cf5fc668186f4159`

**Reply by Nebula-Ninjas** [reply]
@joshlarsen 

I pulled the new engine and wiped and re-pulled the entire repo to build from scratch and now getting the following when running docker-compose up:

opencspm_core_1 exited with code 7
opencspm_worker_1 exited with code 0
opencspm_worker_1 exited with code 7
worker_1  | Could not find bootsnap-1.7.3 in any of the sources
worker_1  | Run `bundle install` to install missing gems.

It just keeps attempting to run the 'bundle install' command until i force quit. Hopefully this is another easy fix.

Thanks in advance,

William

**Reply by joshlarsen** [reply]
This can happen when switching versions. Try to delete the volumes with:

```
docker-compose down -v
```

**Reply by Nebula-Ninjas** [reply]
@joshlarsen that did the trick. Testing if the custom json file loads and will let you all know if I run into anything.

Thanks for the help and quick turn around on all this!

William

**Reply by Nebula-Ninjas** [reply]
@joshlarsen just giving an update. I deleted the volumes, repulled the repo, and now receiving the following errors when I do a docker-compose up _(see attached file)_.

Thanks in advance,

William

[opencspm_error1.txt](https://github.com/OpenCSPM/opencspm/files/6353672/opencspm_error1.txt)

**Reply by joshlarsen** [reply]
William,

It looks like the Docker volume that contains the database is either not mounted, or is corrupt. You may need to do `docker-compose down -v --remove-orphans`.

There are no new changes in the GitHub repo itself, so you don't need to re-pull that. Just make sure the `0.1.20` tagged image has the right sha256 hash (above) before you `docker-compose up` again.

Hope that helps.

**Reply by Nebula-Ninjas** [reply]
@joshlarsen that helped. Got it all back up and running and now getting the following errors when loading the custom data. I have verified that I am running 0.1.20 for the engine. Hope this is another easy fix.

Thanks in advance!

William

[opencspm_error2.txt](https://github.com/OpenCSPM/opencspm/files/6354054/opencspm_error2.txt)

**Reply by joshlarsen** [reply]
Hi William,

This is the same as your first error. You are hitting some services that we haven't written loaders for yet. In this case, `CodePipeline`. Thank you for helping surface these. Go ahead and follow the same process as before.

Docker compose down and remove volumes:

```
docker-compose down -v --remove-orphans
```

Re-pull image

```
docker pull gcr.io/opencspm/engine:0.1.20
```

Re-up

```
docker-compose up
```

**Reply by Nebula-Ninjas** [reply]
@joshlarsen will do. I'll give an update later tonight or tomorrow morning. Glad I'm providing some helpful info and I greatly appreciate the assistance.

I would like to reiterate that this is one of the coolest open-source tool sets I have seen and relatively easy to get working compared to others. Appreciate all the contributors creating this!

Sincerely,

William

**Reply by Nebula-Ninjas** [reply]
@joshlarsen getting a new one now that is needed _(see error below)_ I thought it was as simple as just creating the *.rb file with the new name _(copying the naming style you guys are using and the code snippet from the other one recently added)_, adding the new AWSLoader with the error name and bringing the containers down / up again. I presume this isn't the case. Is it relatively easy to update locally and provide you with all the ones I had to add so you all can update? Just throwing it out there, I'm game to help. If not, no biggie.

worker_1  | Loader job failed - NameError uninitialized constant AWSLoader::CodeBuild (/app/app/jobs/lib/loader/asset/loaders/aws_loader/_resource_loader.rb:25:in `const_get')
worker_1  | Runner job failed - NameError uninitialized constant AWSLoader::CodeBuild (/app/app/jobs/lib/loader/asset/loaders/aws_loader/_resource_loader.rb:25:in `const_get'

Sincerely,

William

**Reply by joshlarsen** [reply]
Hi William,

Appreciate the offer to help. The method you mentioned would work, but you'd have to do it from within the `worker` container and not the local Git source.

I've added `CodeBuild` to the image. Feel free to retry as above. Let me know if others are missing.

**Reply by Nebula-Ninjas** [reply]
Thanks for the direction @joshlarsen I'll give that a try on future ones, below is the next one I am getting:

worker_1  | Loader job failed - NameError uninitialized constant AWSLoader::Redshift

Sincerely,

William

**Reply by Nebula-Ninjas** [reply]
@joshlarsen I added the following to the worker and now getting the below error:

Redshift
ApiGatewayV2

New error:

worker_1  | Loader job failed - FastJsonparser::ParseError A string is opened, but never closed. (/gems/gems/fast_jsonparser-0.5.0/lib/fast_jsonparser.rb:15:in `_parse')
worker_1  | Runner job failed - FastJsonparser::ParseError A string is opened, but never closed. (/gems/gems/fast_jsonparser-0.5.0/lib/fast_jsonparser.rb:15:in `_parse')

Sincerely,

William

**Reply by joshlarsen** [reply]
Hi William,

I've added those to the image now as well. Please re-pull the image and try again. If that issue is still there, we may need to look at the JSON file you are importing.

**Reply by Nebula-Ninjas** [reply]
Thanks @joshlarsen unfortunately I am still getting the error. I've pasted the error that occurs just before the one I sent earlier.

worker_1  | Warning: no EFS loader defined for asset type: filesystem
worker_1  | Loader job failed - FastJsonparser::ParseError A string is opened, but never closed. (/gems/gems/fast_jsonparser-0.5.0/lib/fast_jsonparser.rb:15:in `_parse')
worker_1  | Runner job failed - FastJsonparser::ParseError A string is opened, but never closed. (/gems/gems/fast_jsonparser-0.5.0/lib/fast_jsonparser.rb:15:in `_parse')

Let me know if you'd like to connect / your availability and we can get something on the books.

I'll see if I can find the EFS / filesystem portion and remove either the EFS / the following JSON after it and see if that does the trick.

Thanks in advance,

William

**Reply by joshlarsen** [reply]
The warning on the prior line is ok. Unrelated to this issue.

If you are able to share your JSON with me, feel free to join our [Slack workspace](https://join.slack.com/t/opencspm/shared_invite/zt-pc9bemva-A9rCxkF2FuTbQFd2OlcABw) and DM me directly.

**Reply by Nebula-Ninjas** [reply]
@joshlarsen I will need to redact a lot of information within my JSON file before doing so, so that will take awhile.

Is there a limit to how big a JSON file can be and how many lines a JSON file can have with the OpenCSPM tool?

Sincerely,

William

**Reply by joshlarsen** [reply]
Hi William,

No, there is no size limit. Before you go through the trouble of redacting, you could try to see if the JSON is malformed somewhere. If you have `jq`, could do something like `cat data.json | jq` and try to see what line is having issues.

**Reply by Nebula-Ninjas** [reply]
@joshlarsen just wanted to give you an update. Great call on checking the JSON file, I found some very strange oddities on the originally JSON file like missing parts of words, quotations, and brackets. I attempted to fix it and although it passes the jq test I get the below error, which I know it has to still be related to the JSON file:

worker_1  | Loader job failed - FastJsonparser::ParseError The JSON document has an improper structure: missing or superfluous commas, braces, missing keys, etc. (/gems/gems/fast_jsonparser-0.5.0/lib/fast_jsonparser.rb:15:in `_parse')

Tomorrow my plan is to re-run the aws_recon / aws-vault combo and re-generate the JSON file, verify / fix / remove bad lines all together and re-attempt loading it into opencspm. I will report back as soon as I have results.

William

**Reply by joshlarsen** [reply]
Let me know the flags you are running `aws_recon` with. If the output file is large, you may have to use the file output options as opposed to redirecting stdout.

**Reply by Nebula-Ninjas** [reply]
AWS_PROFILE=coda-dev /gems/bin/aws_recon -j \
-r global,us-east-1 \
-f custom > output.json

Posting a potential issue on the aws_recon repo now as well as I think I may have found another issue but will let you confirm.

Sincerely,

William

**Reply by Nebula-Ninjas** [reply]
@joshlarsen giving an update:

I was finally able to get the JSON file to load after removing some services that are not fully supported _(lambda's w/ vpc_id, DirectoryService)_. I'll post to my other thread on my experiences with aws_recon in hopes it will help.

Last question I have _(for the time being..LOL)_ is do you guys have an option to export to CSV? I have not checked and will do so in a bit, just wanted to ask while it was fresh on my brain.

Thank you for all the direction and assistance and let me know if you need me to test anything out for you all. More than happy to do so.

**Reply by joshlarsen** [reply]
Hi William,

There is no CSV export option as of yet. Where would you want to see that? In the controls list view?

**Reply by Nebula-Ninjas** [reply]
@joshlarsen where ever it would be the easiest to add the option. It would be a bonus if when exported it listed the issues with the severity rating.

Thanks in advance,

William
